Removing the need for always rebuild

tejlmand opened the pull request:
This commit removes the need for always rebuild when generating the
syscalls.json.

It find the files and subfolders of the include directory and makes
syscalls depends on those.
To find added files / subfolders on all platforms CMake configure
is also depending on the folders found.
This ensures that added files / folders will also be detected at later
stages.

Fixes: #8210

tejlmand committed:
cmake: Removing the need for always rebuild

Fixes: #8210

This commit removes the need for always rebuild when generating the
syscalls.json.

It find the files and subfolders of the include directory and makes
syscalls depends on those.
To find added files / subfolders on all platforms CMake configure
is also depending on the folders found.
This ensures that added files / folders will also be detected at later
stages.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

SebastianBoe commented:
`set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${PARSE_SYSCALLS_PATHS_DEPENDS})
`

Having CMake re-run whenever a header file in include/ is modified is a significant regression wrt. incremental build-time performance.

I wonder if there is a way to have parse-syscalls depend on the files, and have cmake reconfiguration depend on the directories. That would have acceptable performance.

tejlmand commented:
If you notice.
The loop will fetch only the directories and populate PARSE_SYSCALLS_PATHS_DEPENDS with only the folders.
So the PR does tell CMake configure to only depend on the directories.
(That's why that variable is called PATHS).
But that means we get the penalty hits on OS'es where a modification to a file is also seen as a change to the folder.

tejlmand commented:
if a manual run of cmake is acceptable when adding a new sub-folder, then we can limit:
set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${PARSE_SYSCALLS_PATHS_DEPENDS})
to windows hosts.
That way, on Linux, adding new files (but not new sub-folders) can be detected directly on the DEPENDS on custom command.
On windows, the cmake re-configure will only trigger when files are added / removed and not updated

SebastianBoe commented:
Ah, my mistake, I misread the code.

"But that means we get the penalty hits on OS'es where a modification to a file is also seen as a change to the folder."

Hmm, and this exists? Is it a soon-to-be deprecated platform? e.g. win7?

If not then we need to rethink this.

Is it possible to have a dependency chain like this?

folder-change -> new_folder?.py -> cmake reconfigure

SebastianBoe commented:
"if a manual run of cmake is acceptable when adding a new sub-folder"
Manual? No. That will cause a user-error pitfall when a user copies a folder with a header file into include/. But automatic is OK.

tejlmand commented:
Manual = No, was also my assumption.
Our last problem here, is that we only want to detect added / removed files to trigger re-run of cmake.

tejlmand committed:
cmake: Update to dependency handling for syscalls.json

Fixes: #8210

Following changes has been made to ensure correct behavior on different
system:
- Python script to detect changes to directories, including empty ones.
  When files are modified the list is updated
  If sub-directories are added / removed a trigger file is touched to
  notify cmake to re-run
- Windows: To detect changes to header files in include for
           parse_syscalls.py all files must be individual monitored.
           Hence all headers are globbed added to dependencies.
           CMake configure depends on the folders so the added /
           removed files are picked up.
- Other:   Folders are monitored through the python list file so that
           added / remove / modified
           Added / removed sub-directories are detected through trigger
           file in order to re-run cmake.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand commented:
A new update with the following changes.

Main point is, that windows / linux behave differently and thus needs slightly different handling.

Windows needs to monitor each file which can easily be done using file globbing and populate a list.
If a file/folder is later added or removed the glob list must be updated, i.e. cmake configure must run.
This is handled by setting a dependency on the directory for cmake configure.
To ensure empty folders are also correctly monitored, a python script has been added.

Linux/other: A modification to a file will be reflected as a folder change, hence, monitoring folders in same fashion as on windows would trigger to often cmake configure invocations.
Instead the python script is used to touch a trigger file whenever a folder is added / removed, and thus cmake configure will be invoked.
On linux it is enough to monitor folder modifications as all file modifications are also reflected on folder level. Hence, no need to monitor each header file. This also means that adding / removing a header file is directly detected in ninja / make and no need to invoke cmake in those case.
This limits cmake invocation to only occur when folders are added / removed.

SebastianBoe commented:
This sounds great!

Just what we need :)

SebastianBoe commented:
This was very verbose, is there anything we can do about it?

`[1/92] Generating ../../../../include, ../../../../include/dfu, ../../../../include/usb, ../../../../include/toolchain, ../../../../include/crypto, ../../../../include/logging, ../../../../include/misc, ../../../../include/debug, ../../../../include/drivers, ../../../../include/mgmt, ../../../../include/shell, ../../../../include/settings, ../../../../include/dt-bindings, ../../../../include/arch, ../../../../include/display, ../../../../include/linker, ../../../../include/fs, ../../../../include/net, ../../../../include/posix, ../../../../include/bluetooth, ../../../../include/nvs, ../../../../include/zephyr, ../../../../include/random, ../../../../include/usb/class, ../../../../include/drivers/usb, ../../../../include/drivers/pci, ../../../../include/drivers/ieee802154, ../../../../include/drivers/gpio, ../../../../include/drivers/display, ../../../../include/drivers/wifi, ../../../../include/drivers/entropy, ../../../../include/drivers/console, ../../../../include/drivers/bluetooth, ../../../../include/drivers/clock_control, ../../../../include/dt-bindings/clock, ../../../../include/dt-bindings/interrupt-controller, ../../../../include/dt-bindings/pinctrl, ../../../../include/dt-bindings/gpio, ../../../../include/dt-bindings/i2c, ../../../../include/arch/arm, ../../../../include/arch/riscv32, ../../../../include/arch/x86, ../../../../include/arch/arc, ../../../../include/arch/posix, ../../../../include/arch/nios2, ../../../../include/arch/xtensa, ../../../../include/arch/arm/cortex_m, ../../../../include/arch/arm/cortex_m/mpu, ../../../../include/arch/arm/cortex_m/scripts, ../../../../include/arch/riscv32/fe310, ../../../../include/arch/riscv32/common, ../../../../include/arch/riscv32/riscv32-qemu, ../../../../include/arch/riscv32/pulpino, ../../../../include/arch/arc/v2, ../../../../include/arch/arc/v2/mpu, ../../../../include/posix/sys, ../../../../include/bluetooth/mesh
`

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -332,12 +332,62 @@ add_custom_command(                       OUTPUT  ${syscall_macros_h}
+set(syscall_list_h   ${CMAKE_CURRENT_BINARY_DIR}/include/generated/syscall_list.h)
+set(syscalls_json    ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls.json)
+set(syscalls_subdirs ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls_subdirs.txt)
+set(syscalls_trigger ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls_subdirs.trigger)
```
Could you add documentation somewhere that explains why syscalls_subdirs.txt could not be used as the trigger file?

tejlmand replied:
will do.

Currently I can inform that the reason is the difference between Linux and windows.
On windows, cmake will be re-invoked just when folder is updated.
However, on Linux we need to differentiate  between a modified file in existing folder (this triggers subdir to have its timestamp updated, and thus json re-run), whereas an added / removed folder triggers the trigger file, which causes cmake to re-run.

It is to aoid havingwith having two custom_commands depending on ${PARSE_SYSCALLS_PATHS_DEPENDS} which can make other strange behavior.
An option can be to make an add_custom_target to which multiple other targets can depend

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -332,12 +332,62 @@ add_custom_command(                       OUTPUT  ${syscall_macros_h}
-set(syscalls_json  ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls.json)
+set(syscall_list_h   ${CMAKE_CURRENT_BINARY_DIR}/include/generated/syscall_list.h)
+set(syscalls_json    ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls.json)
+set(syscalls_subdirs ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls_subdirs.txt)
```
to be consistent with the two above variables this should be

syscalls_subdirs_txt

same with the below variable.

tejlmand replied:
will do.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -332,12 +332,62 @@ add_custom_command(                       OUTPUT  ${syscall_macros_h}
+  --out-file         ${syscalls_subdirs}           # Write this file
+  )
+file(READ ${syscalls_subdirs} PARSE_SYSCALLS_PATHS_DEPENDS)
+string(REGEX REPLACE "\n" ";" PARSE_SYSCALLS_PATHS_DEPENDS ${PARSE_SYSCALLS_PATHS_DEPENDS})
```
Variables should be treated as if they were immutable whenever possible.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -332,12 +332,62 @@ add_custom_command(                       OUTPUT  ${syscall_macros_h}
+  --out-file         ${syscalls_subdirs}           # Write this file
+  )
+file(READ ${syscalls_subdirs} PARSE_SYSCALLS_PATHS_DEPENDS)
+string(REGEX REPLACE "\n" ";" PARSE_SYSCALLS_PATHS_DEPENDS ${PARSE_SYSCALLS_PATHS_DEPENDS})
```
I think these two invocations are equivalent to

  file(
    STRINGS
    ${syscalls_subdirs}
    PARSE_SYSCALLS_PATHS_DEPENDS
  )

tejlmand replied:
seems you're right.
Will make code cleaner.

SebastianBoe commented:
It seems that the "someone copies a folder of syscalls into include" use-case doesn't work well.

```
sebo@mach:~/zephyr/samples/hello_world/b$ cp -r ~/zephyr/include/drivers/ ~/zephyr/include/drivers2/ 
sebo@mach:~/zephyr/samples/hello_world/b$ ninja
[1/98] Generating misc/generated/syscalls_subdirs.txt
Touching
[2/98] Generating misc/generated/syscalls.json
sebo@mach:~/zephyr/samples/hello_world/b$ ninja
[0/1] Re-running CMake...
-- Selected BOARD nrf52840_pca10056
Zephyr version: 1.12.0
Parsing Kconfig tree in /home/sebo/zephyr/Kconfig
Using /home/sebo/zephyr/samples/hello_world/b/zephyr/.config as base
-- Generating zephyr/include/generated/generated_dts_board.h
extract_dts_includes.py: Merge of 'node_type': '['USB-EP', 'USB']'  overwrites '['USB']'.
extract_dts_includes.py: Merge of 'title': 'Nordic nRF52 USBD'  overwrites 'USB Base Structure'.
extract_dts_includes.py: Merge of 'description': 'This binding gives a base representation of the Nordic nRF52 USB device controller
'  overwrites 'This binding gives the base structures for all USB devices
'.
-- Cache files will be written to: /home/sebo/.cache/zephyr
-- Configuring done
-- Generating done
-- Build files have been written to: /home/sebo/zephyr/samples/hello_world/b
[1/128] Generating ../../../../include, ../../../../include/bufaroe, ../../../../include/dfu, ../../../../include/usb, ../../../../include/toolchain, ../../../../include/crypto, ../../../../include/logging, ../../../../include/misc, ../../../../include/debug, ../../../../include/drivers, ../../../../include/mgmt, ../../../../include/shell, ../../../../include/settings, ../../../../include/dt-bindings, ../../../../include/figaro, ../../../../include/arch, ../../../../include/display, ../../../../include/drivers2, ../../../../include/linker, ../../../../include/fs, ../../../../include/net, ../../../../include/posix, ../../../../include/bluetooth, ../../../../include/nvs, ../../../../include/zephyr, ../../../../include/random, ../../../../include/usb/class, ../../../../include/drivers/usb, ../../../../include/drivers/pci, ../../../../include/drivers/ieee802154, ../../../../include/drivers/gpio, ../../../../include/drivers/display, ../../../../include/drivers/wifi, ../../../../include/drivers/entropy, ../../../../include/drivers/console, ../../../../include/drivers/bluetooth, ../../../../include/drivers/clock_control, ../../../../include/dt-bindings/clock, ../../../../include/dt-bindings/interrupt-controller, ../../../../include/dt-bindings/pinctrl, ../../../../include/dt-bindings/gpio, ../../../../include/dt-bindings/i2c, ../../../../include/arch/arm, ../../../../include/arch/riscv32, ../../../../include/arch/x86, ../../../../include/arch/arc, ../../../../include/arch/posix, ../../../../include/arch/nios2, ../../../../include/arch/xtensa, ../../../../include/arch/arm/cortex_m, ../../../../include/arch/arm/cortex_m/mpu, ../../../../include/arch/arm/cortex_m/scripts, ../../../../include/arch/riscv32/fe310, ../../../../include/arch/riscv32/common, ../../../../include/arch/riscv32/riscv32-qemu, ../../../../include/arch/riscv32/pulpino, ../../../../include/arch/arc/v2, ../../../../include/arch/arc/v2/mpu, ../../../../include/drivers2/usb, ../../../../include/drivers2/pci, ../../../../include/drivers2/ieee802154, ../../../../include/drivers2/gpio, ../../../../include/drivers2/display, ../../../../include/drivers2/wifi, ../../../../include/drivers2/entropy, ../../../../include/drivers2/console, ../../../../include/drivers2/bluetooth, ../../../../include/drivers2/clock_control, ../../../../include/posix/sys, ../../../../include/bluetooth/mesh
Preparing syscall parsing
[2/127] Generating misc/generated/syscalls.json
[3/126] Building C object zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj

```

I guess it is impossible to have the build system touch a file that triggers a CMake-rerun in the same invocation. Hmm, it is not good that the user needs to invoke ninja twice.

Is there nothing we can do here?

SebastianBoe commented on scripts/subfolder_list.py:
```diff
@@ -0,0 +1,58 @@
+    if(trigger is None):
+        return
+    if os.path.exists(trigger):
+        print("Touching")
```
This needs to either be removed, or be more verbose.

tejlmand replied:
actually the print message was for debugging, and should have been removed.
Will add comment on why to retrurn.

tejlmand replied:
It is not necessary to run ninja twice.
The first invocation will detect the new files and sys call are parsed correctly.
However, before next invocation of ninja, we need to ensure the header dependencies are updated in case they will be modified.
So therefore CMake is invoked on second ninja.

SebastianBoe replied:
"It is not necessary to run ninja twice."

My mistake.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -332,12 +332,62 @@ add_custom_command(                       OUTPUT  ${syscall_macros_h}
+  # This command sets up the target for CMake to find.
+  add_custom_command(OUTPUT ${PARSE_SYSCALLS_PATHS_DEPENDS}
+    COMMAND ${CMAKE_COMMAND} -E echo "Preparing syscall parsing"
+    )
```
I am not able to understand why this is needed, things appeared to work when I removed it.

tejlmand replied:
Did you try to delete a folders inside the include directory ?

tejlmand replied:
Did you try to delete a folders inside the include directory ?

The reason is that in case the build depend on a folder, let's say:
include/MyFolder, 
then the depends tells CMake that a include/MyFolder exists to depend upon.
If that folder is removed, then CMake (ninja) will try to find the target which has the output named: include/MyFolder which it cannot find, and thus it will fail.

adding a custom command specifyinf this as output gives CMake a target to execute when folder is removed.

SebastianBoe replied:
Thank you for the explanation.

I understand how it works now.

We need to depend on the directories, because we need to know when a folder has been created/deleted.

But if we depend on a directory and it get's deleted the build system will report that there is a missing dependency and no way to create it. So we lie to the build system and claim that the directories can be created if a certain command is run.

SebastianBoe commented on scripts/subfolder_list.py:
```diff
@@ -0,0 +1,58 @@
+            dirlist.extend(os.path.join(root, subdir))
+            dirlist.extend(os.linesep)
+
+    ofile = args.out_file
```
this variable name is unfortunate.

perhaps either directly use args.out_file

or use 'out_file' so we don't have two names for the same thing.

tejlmand replied:
point taken.
will change to out_file

SebastianBoe reviewed (changes requested):
Feedback inline.

SebastianBoe commented:
This looks good. Will approve when all review comments have been addressed.

tejlmand commented:
All comments addressed.

SebastianBoe commented:
@nashif, @carlescufi : This is ready to be merged.

The main feature here is that no-action builds now behave like this:

```
sebo@mach:~/zephyr/samples/hello_world/b$ ninja
ninja: no work to do.

```

instead of

```
sebo@mach:~/zephyr/samples/hello_world/b$ ninja
[1/98] Generating always_rebuild
Building for board nrf52840_pca10056

```
